extensions/machine-learning-services/src/controllers/mainController.ts

```diff
@@ -0,0 +1,73 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as vscode from 'vscode';
+import * as nbExtensionApis from '../typings/notebookServices';
+import { isNullOrUndefined } from 'util';
+import { PackageManager } from '../packageManagement/packageManager';
+import * as constants from '../common/constants';
+import { ApiWrapper } from '../common/apiWrapper';
+import { QueryRunner } from '../common/queryRunner';
+
+const NotebookExtensionName = 'Microsoft.notebook';
+
+/**
+ * The main controller class that initializes the extension
+ */
+export default class MainController implements vscode.Disposable {
+	protected _context: vscode.ExtensionContext;
+	private _nbExtensionApis: nbExtensionApis.IExtensionApi;
+	private _packageManager: PackageManager;
+	private _outputChannel: vscode.OutputChannel;
+	private _rootPath: string;
+	private _apiWrapper: ApiWrapper;
+	private _queryRunner: QueryRunner;
+
+	public constructor(context: vscode.ExtensionContext) {
+		this._context = context;
+		this._outputChannel = vscode.window.createOutputChannel(constants.extensionOutputChannel);
+	}
+
+	public deactivate(): void {
+	}
+
+	public activate(): Promise<boolean> {
+		this.initialize();
+		return Promise.resolve(true);
+	}
+
+	/**
+	 * Returns an instance of Jupyter Server Installation from notebook extension
+	 */
+	private getNotebookExtensionApis(): nbExtensionApis.IExtensionApi {
+		if (isNullOrUndefined(this._nbExtensionApis)) {
+			// TODO: wait for extension to load
```

Comment:
Will there be a task for this?

Reply:
do you know how I can do this? I can fix it in this PR or create a task. but cannot find how to do it correctly 

---
In reply to: [356367736](https://github.com/microsoft/azuredatastudio/pull/8622#discussion_r356367736) [](ancestors = 356367736)

Reply:
No idea, just wanted to make sure we're tracking TODOs on github. Random comments tend to get forgotten. I think we do some checking for extension startups in the extension unit tests.

Reply:
Use [Extension.activate()](https://code.visualstudio.com/api/references/vscode-api#Extension%3CT%3E) and await on that. (You get the extension by calling [vscode.extensions.getExtension](https://code.visualstudio.com/api/references/vscode-api#extensions.getExtension))

If the extension is already activated it'll just return the API directly.